lib/dynamoid/adapter_plugin/aws_sdk_v2.rb

```diff
@@ -559,6 +569,15 @@ def scan(table_name, scan_hash, select_opts = {})
           record_count = 0
           scan_count = 0
           loop do
+            # Adjust the limit down if the remaining record or scan limit are
+            # lower to obey limits. We can assume the difference won't be
+            # negative due to break statements below.
+            if request[:limit] && record_limit && record_limit - record_count < request[:limit]
+              request[:limit] = record_limit - record_count
+            elsif request[:limit] && scan_limit && scan_limit - scan_count < request[:limit]
+              request[:limit] = scan_limit - scan_count
+            end
+
```

Comment:
There is a case when this approach works not well. I don't know how to solve the issue now but we should remember  that it exists.

We tune limit according to already loaded items count and can appear in a situation when `request[:limit]` is very small in comparison to items count we need to scan to load all requested items (in case when user specified some conditions) and we will make a lot of excessive HTTP calls. For instance `request[:limit]` is 1 but nearest matched item is at 100th position. So we will make 100 HTTP calls.

Reply:
> Hm, I don't see the statement you've quoted. Most statements can be interpreted as the resulting set has maximum of 1 MB. Nothing about the amount of data that is looked at in underlying query/scan.

Just FYI - there is a clean statement in provided article:

"A Query operation can retrieve a maximum of 1 MB of data. This limit applies before the filter expression is evaluated."